### PR TITLE
Delegation Pattern

### PR DESCRIPTION
**Pull request by mikulucky**
Delegation Pattern Implemented as requested by issue #324 should also be noted unit tests have been provided so updates required to #293 by adding and checking it off.

### CODE REVIEW

**Comment by iluwatar** [comment]
The unit tests in `DelegateTest` fail when run under Windows:
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.iluwatar.delegation.simple.AppTest
HP Printer : hello world
Canon Printer : hello world
Epson Printer : hello world
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.016 sec - in com.iluwatar.delegation.simple.AppTest
Running com.iluwatar.delegation.simple.DelegateTest
Canon Printer : Test Message Printed
HP Printer : Test Message Printed
Epson Printer : Test Message Printed
Tests run: 3, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 0.015 sec <<< FAILURE! - in com.iluwatar.delegation.simp
le.DelegateTest
testCanonPrinter(com.iluwatar.delegation.simple.DelegateTest)  Time elapsed: 0.015 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<...Test Message Printed[]
] but was:<...Test Message Printed[
>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at org.junit.Assert.assertEquals(Assert.java:144)
        at com.iluwatar.delegation.simple.DelegateTest.testCanonPrinter(DelegateTest.java:24)

testHpPrinter(com.iluwatar.delegation.simple.DelegateTest)  Time elapsed: 0 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<...Test Message Printed[]
] but was:<...Test Message Printed[
>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at org.junit.Assert.assertEquals(Assert.java:144)
        at com.iluwatar.delegation.simple.DelegateTest.testHpPrinter(DelegateTest.java:32)

testEpsonPrinter(com.iluwatar.delegation.simple.DelegateTest)  Time elapsed: 0 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<...Test Message Printed[]
] but was:<...Test Message Printed[
>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at org.junit.Assert.assertEquals(Assert.java:144)
        at com.iluwatar.delegation.simple.DelegateTest.testEpsonPrinter(DelegateTest.java:40)


Results :

Failed tests:
  DelegateTest.testCanonPrinter:24 expected:<...Test Message Printed[]
] but was:<...Test Message Printed[
>
  DelegateTest.testEpsonPrinter:40 expected:<...Test Message Printed[]
] but was:<...Test Message Printed[
>
  DelegateTest.testHpPrinter:32 expected:<...Test Message Printed[]
] but was:<...Test Message Printed[
>

Tests run: 4, Failures: 3, Errors: 0, Skipped: 0


Might be because of different line ending between Linux and Windows.

File: delegation/index.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+title: Delegation
+folder: delegation
+permalink: /patterns/delegation/
+categories: Behavioral
+tags:
+ - Java
+ - Difficulty-Beginner
+---
+
+**Also known as:** Proxy Chains
```
Comment (iluwatar) [inline delegation/index.md]: Are you sure about this alias name?

File: delegation/index.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+![alt text](./etc/delegation.png "Delegate")
+
+**Applicability:** Use the Delegate pattern in order to achieve the following
+
+* Reduce the coupling of methods to their class
+* Components that behave identically, but realize that this situation can change in the future.
+
+**Credits**
+
+* [Design Patterns: Elements of Reusable Object-Oriented Software](http://www.amazon.com/Design-Patterns-Elements-Reusable-Object-Oriented/dp/0201633612)
```
Comment (iluwatar) [inline delegation/index.md]: I think Delegate pattern is not in this book.

File: delegation/pom.xml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  ~ all copies or substantial portions of the Software.
+  ~
+  ~ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+  ~ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  ~ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  ~ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  ~ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  ~ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+  ~ THE SOFTWARE.
+  -->
```
Comment (iluwatar) [inline delegation/pom.xml]: Remove the license text here, for consistency.

File: delegation/pom.xml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.github.stefanbirkner</groupId>
+            <artifactId>system-rules</artifactId>
+            <version>1.14.0</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (iluwatar) [inline delegation/pom.xml]: The dependencies should be declared in the main `pom.xml`s dependency management section. The dependency should then be referred here without the version number. For an example, check how the above `junit` dependency works.

File: delegation/src/main/java/com/iluwatar/delegation/simple/App.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import com.iluwatar.delegation.simple.printers.EpsonPrinter;
+import com.iluwatar.delegation.simple.printers.HpPrinter;
+
+/**
+ * In this example the delegates are {@link EpsonPrinter}, {@link HpPrinter} and {@link CanonPrinter} they all implement
+ * {@link Printer}. The {@link AbstractPrinterController} and through inheritance {@link PrinterController} also
+ * implement {@link Printer}. However neither provide the functionality of {@link Printer} by printing to the screen,
+ * they actually call upon the instance of {@link Printer} that they were instantiated with. Therefore delegating the
+ * behaviour to another class.
+ */
```
Comment (iluwatar) [inline delegation/src/main/java/com/iluwatar/delegation/simple/App.java]: Please also explain generally what delegation means. I think this should be at the top.

File: delegation/src/main/java/com/iluwatar/delegation/simple/AbstractPrinterController.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * Extra layer of abstraction for the controller to allow the controller in this example {@link PrinterController} to
+ * be as clean as possible. This just provides the default constructor and a simple getter method. The generic of
+ * T allows any implementation of {@link Printer}
+ *
+ * @param <T> Printer
+ * @see Printer
+ * @see PrinterController
+ */
+public abstract class AbstractPrinterController<T extends Printer> implements Printer {
```
Comment (iluwatar) [inline delegation/src/main/java/com/iluwatar/delegation/simple/AbstractPrinterController.java]: Type parameters don't bring any extra value here - you don't even use them. Remove generics and replace `T` with `Printer`.

File: delegation/src/main/java/com/iluwatar/delegation/simple/AbstractPrinterController.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+
+  private T printer;
```
Comment (iluwatar) [inline delegation/src/main/java/com/iluwatar/delegation/simple/AbstractPrinterController.java]: final

File: delegation/src/main/java/com/iluwatar/delegation/simple/PrinterController.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package com.iluwatar.delegation.simple;
+
+public class PrinterController extends AbstractPrinterController {
+
+  public PrinterController(Printer printer) {
```
Comment (iluwatar) [inline delegation/src/main/java/com/iluwatar/delegation/simple/PrinterController.java]: In my opinion there is one layer too many of abstraction. I would remove `AbstractPrinterController` and do everything in `PrinterController`.

**Comment by iluwatar** [comment]
@mikulucky you have my review remarks. Please comment on this thread once ready for another review.

File: delegation/pom.xml (hunk 3 above)
Comment (mikulucky) [inline delegation/pom.xml]: So sorry for this, overhang configuration from another project.

File: delegation/index.md (hunk 1 above)
Comment (mikulucky) [inline delegation/index.md]: They are very similar https://en.wikipedia.org/wiki/Proxy_pattern might be better to rename the aka to Proxy Pattern

**Comment by mikulucky** [comment]
I believe I have addressed all the review comments, when you able to could you pleased resume the review please. Thanks. :smile:

**Comment by iluwatar** [comment]
@mikulucky looks good. Thank you for the contribution :smile:

**Pull request merged**